pandas/core/indexes/base.py

```diff
@@ -4720,6 +4720,10 @@ def get_indexer_non_unique(self, target):
             return pself.get_indexer_non_unique(ptarget)
 
         if self.is_all_dates:
+            # If `target` doesn't consist of dates, `target.asi8` will return
+            # None, which will raise TypeError. GH 27994
+            if not target.is_all_dates:
```

Comment:
this likely is not necessary if you implemented as per my comment above

Reply:
When passing `str`, for example, to dataframe with `DatetimeIndex`, `type(self) == type(target)` would return `False`, in which case I believe `KeyError` should be raised. But underlying arrays aren't empty.
If not here, I feel like Cython code needs changing though I'm not confident.

Reply:
What exception do you get without this? `Index.is_all_dates` can be surprisingly expensive.